pkg/healthcheck/healthcheck.go

```diff
@@ -2191,6 +2214,94 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
 	return nil
 }
 
+// Check if there's a pod with the "opaque ports" annotation defined but a
+// service selecting the aforementioned pod doesn't define it
+func (hc *HealthChecker) checkMisconfiguredOpaquePortAnnotations(ctx context.Context, namespace string) error {
+	services, err := hc.GetServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	var errStrings []string
+
+	for _, service := range services {
+		if service.Spec.ClusterIP == "None" {
+			// skip headless services; they're handled differently
+			continue
+		}
+
+		endpoint, err := hc.kubeAPI.CoreV1().Endpoints(namespace).Get(ctx, service.Name, metav1.GetOptions{})
+
+		if err != nil {
+			return err
+		}
+
+		pods := make([]*corev1.Pod, 0)
+		for _, subset := range endpoint.Subsets {
+			for _, addr := range subset.Addresses {
+				if addr.TargetRef != nil && addr.TargetRef.Kind == "Pod" {
+					pod, err := hc.kubeAPI.CoreV1().Pods(namespace).Get(ctx, addr.TargetRef.Name, metav1.GetOptions{})
+					if err != nil {
+						return err
+					}
+					pods = append(pods, pod)
+				}
+			}
+		}
+
+		if mismatch := misconfiguredOpaquePortAnnotationsInService(service, pods); mismatch != nil {
+			errStrings = append(
+				errStrings,
+				mismatch.Error(),
+			)
+		}
+	}
+
+	if len(errStrings) >= 1 {
+		return fmt.Errorf(strings.Join(errStrings, "\n    "))
+	}
+
+	return nil
+}
+
+func misconfiguredOpaquePortAnnotationsInService(service corev1.Service, pods []*corev1.Pod) error {
+	for _, pod := range pods {
+		if err := misconfiguredOpaqueAnnotation(service, pod); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func misconfiguredOpaqueAnnotation(service corev1.Service, pod *corev1.Pod) error {
+	svcAnnotations := service.Annotations
+	podAnnotations := pod.Annotations
+
+	// If the pod has the annotation, check that the service has it as well
+	if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return fmt.Errorf("pod/%s and service/%s have the annotation but values don't match", pod.Name, service.Name)
+			}
+		} else {
+			return fmt.Errorf("pod/%s has the annotation but service/%s doesn't", pod.Name, service.Name)
+		}
+	}
+
+	// If the service has the annotation, check that the pod has it as well
+	if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return fmt.Errorf("pod/%s and service/%s have the annotation but values don't match", pod.Name, service.Name)
```

Comment:
This is redundant with the above, isn't it?

Another potential way to structure this function to eliminate the redundancy would be:

```
svcAnnotation, svcAnnotationOk := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]
podAnnotation, podAnnotationOk := podAnnotations[k8s.ProxyOpaquePortsAnnotation]
if svcAnnotationOk && podAnnotationOk {
// check they match
}
if svcAnnotationOk {
// pod annotation is missing
}
if podAnnotationOk {
// svc annotation is missing
}
```